artifact.go

```diff
@@ -203,67 +213,423 @@ func (m artifactModule) getBuildTestsResultsHandler(c *gin.Context) {
+							" with ID %d in database. Expected TRX/XML format.", summary.ArtifactID, buildID),
+				})
+			return
+		}
+
+		summary.FileName = artifact.FileName
+		summaries = append(summaries, summary)
+		details = append(details, detail...)
+	}
+
+	summaryList := TestResultListSummary{
+		Summaries: summaries,
+		BuildID:   buildID}
+	for _, summary := range summaryList.Summaries {
+		err := m.Database.
+			Create(summary).
+			Error
+		if err != nil {
+			ginutil.WriteDBWriteError(c, err, fmt.Sprintf(
+				"Failed saving test result summary from artifact with ID %d, for build"+
+					" with ID %d in database.",
+				summary.ArtifactID, buildID))
+			return
+		}
+
+		summaryList.Failed += summary.Failed
+		summaryList.Passed += summary.Passed
+		summaryList.Skipped += summary.Skipped
+	}
+
+	summaryList.Total =
+		summaryList.Failed +
+			summaryList.Passed +
+			summaryList.Skipped
+
+	err := m.Database.
+		CreateInBatches(details, 100).
+		Error
+	if err != nil {
+		ginutil.WriteDBWriteError(c, err, fmt.Sprintf(
+			"Failed saving test result details for build with ID %d in database.",
+			buildID))
+	}
+
+	c.JSON(http.StatusOK, summaryList)
+}
+
+// getBuildAllTestResultDetailsHandler godoc
+// @summary Get all test result details for specified build
+// @tags artifact
+// @param buildid path int true "Build ID"
+// @success 200 {object} []TestResultDetail
+// @failure 400 {object} problem.Response "Bad request"
+// @failure 502 {object} problem.Response "Database is unreachable"
+// @router /build/{buildid}/test-result-details [get]
+func (m artifactModule) getBuildAllTestResultDetailsHandler(c *gin.Context) {
+	buildID, ok := ginutil.ParseParamUint(c, buildIDParamName)
+	if !ok {
+		return
+	}
+
+	details := []TestResultDetail{}
+	err := m.Database.
+		Where(&TestResultDetail{BuildID: uint(buildID)}).
+		Find(&details).
+		Error
 
-	form, err := c.MultipartForm()
 	if err != nil {
-		ginutil.WriteMultipartFormReadError(c, err, fmt.Sprintf(
-			"Failed reading multipart-form content from request body when uploading new artifact for build with ID %d.",
+		ginutil.WriteDBReadError(c, err, fmt.Sprintf(
+			"Failed fetching test result details for build with ID %d from database.",
 			buildID))
 		return
 	}
 
+	c.JSON(http.StatusOK, details)
+}
+
+// getBuildTestResultDetailsHandler godoc
+// @summary Get test result details for specified test
+// @tags artifact
+// @param buildid path int true "Build ID"
+// @param artifactId path int true "Artifact ID"
+// @success 200 {object} []TestResultDetail
+// @router /build/{buildid}/test-result-details/{artifactId} [get]
+func (m artifactModule) getBuildTestResultDetailsHandler(c *gin.Context) {
+	buildID, ok := ginutil.ParseParamUint(c, buildIDParamName)
+	if !ok {
+		return
+	}
+
+	artifactID, ok := ginutil.ParseParamUint(c, artifactIDParamName)
+	if !ok {
+		return
+	}
+
+	details := []TestResultDetail{}
+	err := m.Database.
+		Where(&TestResultDetail{BuildID: uint(buildID), ArtifactID: artifactID}).
```

Comment:
Is this cast really needed?

```suggestion
		Where(&TestResultDetail{BuildID: buildID, ArtifactID: artifactID}).
```